src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs

```diff
@@ -386,6 +386,252 @@ IValue CopyAndUpdateUdt(TupleValue originalValue)
+            {
+                while (true)
+                {
+                    var i = s.IndexOf('{', start);
+                    if (i < 0)
+                    {
+                        return (-1, s.Length, -1);
+                    }
+                    else if ((i == start) || (s[i - 1] != '\\'))
+                    {
+                        var j = s.IndexOf('}', i + 1);
+                        if (j < 0)
+                        {
+                            throw new FormatException("Missing } in interpolated string");
+                        }
+                        var n = int.Parse(s[(i + 1)..j]);
+                        return (i, j + 1, n);
+                    }
+                    start = i + 1;
+                }
+            }
+
+            // Creates a string value that needs to be queued for unreferencing.
+            Value CreateConstantString(string s)
+            {
+                // Deal with escape sequences: \{, \\, \n, \r, \t, \". This is not an efficient
+                // way to do this, but it's simple and clear, and strings are uncommon in Q#.
+                var cleanStr = s.Replace("\\{", "{").Replace("\\\\", "\\").Replace("\\n", "\n")
+                    .Replace("\\r", "\r").Replace("\\t", "\t").Replace("\\\"", "\"");
+
+                Value? constantArray = null;
+                if (cleanStr.Length > 0)
+                {
+                    var constantString = sharedState.Context.CreateConstantString(cleanStr, true);
+                    var globalConstant = sharedState.Module.AddGlobal(
+                        constantString.NativeType, true, Linkage.Internal, constantString);
+                    constantArray = sharedState.CurrentBuilder.GetElementPtr(
+                        sharedState.Context.Int8Type.CreateArrayType((uint)cleanStr.Length + 1), // +1 because zero terminated
+                        globalConstant,
+                        new[] { sharedState.Context.CreateConstant(0) });
+                }
+
+                var zeroLengthString = constantArray == null
+                    ? sharedState.Types.DataArrayPointer.GetNullValue()
+                    : sharedState.CurrentBuilder.BitCast(
+                        constantArray,
+                        sharedState.Types.DataArrayPointer);
+
+                var createString = sharedState.GetOrCreateRuntimeFunction(RuntimeLibrary.StringCreate);
+                return sharedState.CurrentBuilder.Call(createString, zeroLengthString);
+            }
+
+            // Creates a new string with reference count 1 that needs to be queued for unreferencing
+            // and contains the concatenation of both values. Both both arguments are unreferenced.
+            Value DoAppend(Value? curr, Value next, bool unreferenceNext = true)
+            {
+                if (curr == null)
+                {
+                    return next;
+                }
```

Comment:
Just in case I want to make sure that nothing else is needed here depending on the value of `unreferenceNext`.

Reply:
No, nothing else is needed, since we return next; basically next takes the place of what would otherwise be a new string and we hence don't want to release it.

Reply:
Actually, while the code is correct for how it is used, the correct behavior for the function is to increase the reference count for next if unreferenceNext is false.